File: smi/lmd.py
#!/usr/bin/env python3
"""lmd.py -- the metric engine behind the Synaptic Mesh Interface.

    from smi.lmd import metric_from_laplacian, mesh_metric

Distance between two on-screen elements is not authored. It is the effective
resistance between them on the graph of what depends on what:

    L+   = pinv(L)
    R_ij = L+_ii + L+_jj - 2 L+_ij
    d_ij = sqrt(R_ij)

Two elements joined by many strong paths sit close together; two joined only by
a long weak chain drift apart. Nobody positions anything.

*** WHAT IS BEING MEASURED ***
A dependency graph inside running software: which live elements determine which
others, and how strongly. J is the strength of one such dependency; d is how far
apart two elements should sit given all of them. This is telemetry on an
information layer. It is not a statement about matter or distance, and the name
is a name -- see SCOPE.md.

*** GLOBAL COUPLING IS A ZOOM, BY CONSTRUCTION ***
For a scalar J > 0, pinv(J*L0) = (1/J) * pinv(L0). So R(J) = R(1)/J and
d = sqrt(R) is proportional to J^-0.5 EXACTLY -- for every pair, on every graph,
at every size. That is elementary linear algebra, and in an interface it is a
property worth having deliberately: a global tension control rescales the whole
picture and changes NOTHING ELSE. It cannot reorder elements, cannot change what
is near what, and cannot alter a value. It is safe to hand to a user.

The corollary matters just as much: because the -0.5 slope comes out on every
graph, measuring it verifies nothing about any particular graph. It is kept as a
smoke test -- it breaks the moment pinv, the Laplacian or the clipping breaks --
and never reported as a result. What carries information here is TOPOLOGY and
LOCAL coupling.

*** THE TWO PLACES pinv WILL QUIETLY LIE TO YOU ***
1. On a DISCONNECTED graph it returns a finite number between components. Two
   elements with no path between them should be infinitely far apart; the
   pseudo-inverse gives them a modest distance and the layout puts unrelated
   things side by side. Components are therefore detected explicitly here.
2. At J -> 0 the whole Laplacian goes to zero, pinv(0) = 0, and every distance
   collapses to ZERO. A completely broken mesh would render as a perfectly
   contracted one. The dead case is caught before the metric is trusted.

Both are guarded below and both are asserted by smi/test_smi.py.
"""
from __future__ import annotations

import math

from dataclasses import dataclass

import jax
import jax.numpy as jnp
import numpy as np

# A layout engine measured to 1e-6 cannot run at float32: on a 100-node ring the
# identity slope reads -0.500003, which is noise mistaken for signal.
jax.config.update("jax_enable_x64", True)

#: below this total edge weight the mesh is treated as dead rather than measured
DEAD_MESH_EPS = 1e-12

#: A coupling below this is DECLARED FADING: still connected, but too weak to
#: show at display precision. Without a name for it, a wire reading 0.00 in a
#: readout while the legend insists "live" is a fourth, unnamed state -- the
#: interface saying connected and the arithmetic saying I move nothing.
FADE_BELOW = 0.01


@jax.jit
def metric_from_laplacian(L):
    """Proper distances d_ij = sqrt(R_ij) from a graph Laplacian. JIT-compiled.

    Vectorised: one pinv and two broadcasts, no Python loop over pairs. This is
    the whole hot path -- everything else in the interface is bookkeeping.
    """
    L_pinv = jnp.linalg.pinv(L)
    diag = jnp.diag(L_pinv)
    R = diag[:, None] + diag[None, :] - 2.0 * L_pinv
    # tiny negative values are floating-point dust, not negative resistance
    return jnp.sqrt(jnp.clip(R, 0.0))


def components(L, tol=1e-12):
    """Which nodes can actually reach which, by walking the adjacency.

    pinv does not know or care that a graph is in pieces. This does.
    Returns an integer label per node; equal labels mean connected.
    """
    A = np.asarray(L, dtype=float).copy()
    np.fill_diagonal(A, 0.0)
    adj = np.abs(A) > tol
    n = adj.shape[0]
    label = np.full(n, -1, dtype=int)
    nxt = 0
    for start in range(n):
        if label[start] != -1:
            continue
        stack, label[start] = [start], nxt
        while stack:
            u = stack.pop()
            for v in np.nonzero(adj[u])[0]:
                if label[v] == -1:
                    label[v] = nxt
                    stack.append(v)
        nxt += 1
    return label


def mesh_metric(L, tol=1e-12):
    """The distance matrix an interface may actually trust.

    Same as metric_from_laplacian, with the two lies removed:
      * pairs in different components come back as inf, not as a small number
      * a mesh with no coupling left at all comes back as all-inf, not all-zero
    """
    L = np.asarray(L, dtype=float)
    n = L.shape[0]
    off = np.abs(L - np.diag(np.diag(L))).sum()
    if off < DEAD_MESH_EPS:
        # every link is gone. The metric would say zero -- maximum contraction --
        # which is the exact opposite of the truth.
        D = np.full((n, n), np.inf)
        np.fill_diagonal(D, 0.0)
        return D, np.arange(n), True

    # np.asarray on a JAX array gives a READ-ONLY view; the diagonal fix below
    # then fails, but only on connected graphs, because the disconnected branch
    # happens to build a fresh array. Copy explicitly.
    D = np.array(metric_from_laplacian(jnp.asarray(L)), dtype=float, copy=True)
    lab = components(L, tol)
    if lab.max() > 0:
        D = np.where(lab[:, None] == lab[None, :], D, np.inf)
    np.fill_diagonal(D, 0.0)
    return D, lab, False


# ----------------------------------------------------------------- graphs ---
def ring_laplacian(n, J=1.0):
    """Weighted 1-D ring. Each node joined to its two neighbours with weight J."""
    if n < 3:
        raise ValueError("a ring needs at least 3 nodes")
    W = np.zeros((n, n))
    for i in range(n):
        W[i, (i + 1) % n] = W[(i + 1) % n, i] = float(J)
    return np.diag(W.sum(1)) - W


def laplacian_from_edges(n, edges):
    """Laplacian from (i, j, weight) triples. Symmetric by construction."""
    W = np.zeros((n, n))
    for i, j, w in edges:
        if i == j:
            raise ValueError(f"a node cannot depend on itself: node {i}")
        W[i, j] = W[j, i] = float(w)
    return np.diag(W.sum(1)) - W


@dataclass(frozen=True)
class Sweep:
    """The result of scaling every edge by J and watching distance contract."""
    couplings: np.ndarray
    distances: np.ndarray
    slope: float
    r_squared: float
    pair: tuple

    @property
    def matches_identity(self):
        return abs(self.slope + 0.5) <= 1e-4 and self.r_squared >= 0.999999


def sweep_coupling(L_unit, pair, lo=-1.0, hi=2.0, steps=15):
    """Scale a unit-weight Laplacian across J and record one pair's distance.

    The slope this produces is -0.5 on ANY graph. It is a check that the engine
    is right, not a discovery about the graph.
    """
    i, j = pair
    Js = np.logspace(lo, hi, steps)
    ds = np.array([float(metric_from_laplacian(jnp.asarray(J * np.asarray(L_unit)))[i, j])
                   for J in Js])
    if np.any(ds <= 0):
        raise ValueError("a zero distance appeared; the pair is probably not connected")
    slope, _ = np.polyfit(np.log10(Js), np.log10(ds), 1)
    r2 = float(np.corrcoef(np.log10(Js), np.log10(ds))[0, 1] ** 2)
    return Sweep(Js, ds, float(slope), r2, (i, j))


def procrustes2d(Q, P):
    """Rotate/reflect a new embedding onto the previous one.

    Classical MDS fixes an embedding only up to rotation and reflection: the
    eigenvectors are defined up to sign, so a small change in the graph can hand
    back the same picture upside down. Deterministic per input, and to a person
    dragging it, the map flipping under their finger reads as the positions
    being arbitrary -- which is exactly the claim this engine makes against.

    Whichever orthogonal transform moves the shared nodes least is chosen. Real
    change still shows; the cosmetic flips stop. Closed form: 2-D needs no SVD.
    """
    Q = np.asarray(Q, dtype=float)
    P = np.asarray(P, dtype=float)
    n = min(len(Q), len(P))
    if n < 2:
        return Q.copy()
    qc, pc = Q[:n].mean(0), P[:n].mean(0)
    best = None
    for flip in (1.0, -1.0):
        q = (Q[:n] - qc) * np.array([flip, 1.0])
        p = P[:n] - pc
        num = float((q[:, 0] * p[:, 1] - q[:, 1] * p[:, 0]).sum())
        den = float((q[:, 0] * p[:, 0] + q[:, 1] * p[:, 1]).sum())
        th = math.atan2(num, den)
        c, s = math.cos(th), math.sin(th)
        rot = np.array([[c, -s], [s, c]])
        resid = float((((q @ rot.T) - p) ** 2).sum())
        if best is None or resid < best[0]:
            best = (resid, th, flip)
    _, th, flip = best
    c, s = math.cos(th), math.sin(th)
    rot = np.array([[c, -s], [s, c]])
    return ((Q - qc) * np.array([flip, 1.0])) @ rot.T + pc


FLAT_WARN = 0.25


def layout2d(D, keep):
    """Classical MDS: double-centre the squared distances, take the top two
    eigenvectors. The flat picture that best preserves the metric.

    Mirrors smi/lmd.js layout2d exactly; smi/test_parity.py checks that.
    """
    keep = list(keep)
    m = len(keep)
    sq = np.array([[float(D[a][b]) ** 2 for b in keep] for a in keep], dtype=float)
    rm = sq.mean(1)
    B = -0.5 * (sq - rm[:, None] - rm[None, :] + rm.mean())
    w, V = np.linalg.eigh(B)
    order = np.argsort(w)[::-1]
    k1 = order[0]
    k2 = order[1] if m > 1 else order[0]
    return np.column_stack([
        V[:, k1] * math.sqrt(max(float(w[k1]), 0.0)),
        V[:, k2] * math.sqrt(max(float(w[k2]), 0.0)),
    ])


def flatness(D, keep, xy):
    """How much of the metric survived being flattened onto a plane.

    Returns (worst_ratio, a, b): the pair whose DRAWN separation is the smallest
    fraction of its TRUE distance, and which pair that is. 1.0 means every
    distance in the picture is the real one.

    This is not a nicety. A mesh of five elements generally needs four
    dimensions; a screen has two, and classical MDS drops the difference in
    silence. On the invoice mesh SMI ships with, `VAT` and `Total` are drawn on
    top of each other while their true distance is 0.5 -- among the largest in
    the mesh. An interface whose whole claim is that POSITION MEANS SOMETHING
    cannot leave that unsaid, any more than it can print 0.00 for a wire it
    still calls live. Measure it, name it, show it.
    """
    keep = list(keep)
    worst, pair = 1.0, (None, None)
    for i in range(len(keep)):
        for j in range(i + 1, len(keep)):
            true_d = float(D[keep[i]][keep[j]])
            if not np.isfinite(true_d) or true_d <= 0:
                continue
            drawn = float(np.hypot(xy[i][0] - xy[j][0], xy[i][1] - xy[j][1]))
            ratio = drawn / true_d
            if ratio < worst:
                worst, pair = ratio, (keep[i], keep[j])
    return worst, pair[0], pair[1]


def normalised(D):
    """Layout SHAPE with scale divided out -- what H1 says J cannot change."""
    finite = D[np.isfinite(D)]
    m = finite.max() if finite.size else 1.0
    return D / (m if m > 0 else 1.0)

File: smi/lmd.js
/* lmd.js -- the LMD metric engine, in the browser.
 * ===========================================================================
 * A faithful port of smi/lmd.py so SMI runs on a phone with no server. A port
 * is a liability unless it is checked: smi/test_parity.py runs both engines
 * over a shared list of graphs and fails if any distance differs by more than
 * 1e-9. The port is not trusted; it is tested.
 *
 * There is no linear-algebra library here on purpose. The Laplacian is real and
 * symmetric, so the cyclic Jacobi eigenvalue algorithm gives an exact
 * eigendecomposition in about forty lines, and the pseudo-inverse is then
 *
 *     pinv(L) = V · diag(1/λ for λ above tolerance, else 0) · Vᵀ
 *
 * which is all this needs. Fifty nodes is well under a millisecond.
 *
 * WHAT IS BEING MEASURED
 * A dependency graph inside running software: which live elements determine
 * which others, and how strongly. J is the strength of a dependency; d is how
 * far apart two elements should sit given all of them. Information layer. Not
 * a claim about matter or physical distance.
 */
(function (root) {
  "use strict";

  var DEAD_MESH_EPS = 1e-12;

  /* A coupling below this is DECLARED FADING: still connected, but too weak to
   * show at display precision. Without a name for it, a wire reading 0.00 in
   * the readout while the legend insists "live" is a fourth, unnamed state --
   * the interface saying connected and the arithmetic saying I move nothing. */
  var FADE_BELOW = 0.01;

  /* Cyclic Jacobi for a real symmetric matrix. Returns { values, vectors }
   * where vectors[i][k] is component i of eigenvector k. */
  function eigSymmetric(Ain, sweeps) {
    var n = Ain.length, i, j, k, p, q, s;
    var A = Ain.map(function (r) { return r.slice(); });
    var V = [];
    for (i = 0; i < n; i++) {
      V.push(new Array(n).fill(0));
      V[i][i] = 1;
    }
    sweeps = sweeps || 60;
    for (s = 0; s < sweeps; s++) {
      var off = 0;
      for (i = 0; i < n; i++) for (j = i + 1; j < n; j++) off += A[i][j] * A[i][j];
      if (off < 1e-30) break;
      for (p = 0; p < n - 1; p++) {
        for (q = p + 1; q < n; q++) {
          if (Math.abs(A[p][q]) < 1e-300) continue;
          var theta = (A[q][q] - A[p][p]) / (2 * A[p][q]);
          var t = Math.sign(theta || 1) / (Math.abs(theta) + Math.sqrt(theta * theta + 1));
          var c = 1 / Math.sqrt(t * t + 1), sn = t * c;
          for (k = 0; k < n; k++) {
            var akp = A[k][p], akq = A[k][q];
            A[k][p] = c * akp - sn * akq;
            A[k][q] = sn * akp + c * akq;
          }
          for (k = 0; k < n; k++) {
            var apk = A[p][k], aqk = A[q][k];
            A[p][k] = c * apk - sn * aqk;
            A[q][k] = sn * apk + c * aqk;
          }
          for (k = 0; k < n; k++) {
            var vkp = V[k][p], vkq = V[k][q];
            V[k][p] = c * vkp - sn * vkq;
            V[k][q] = sn * vkp + c * vkq;
          }
        }
      }
    }
    var vals = [];
    for (i = 0; i < n; i++) vals.push(A[i][i]);
    return { values: vals, vectors: V };
  }

  /* d_ij = sqrt(R_ij), R_ij = L⁺_ii + L⁺_jj − 2·L⁺_ij. Same as the Python. */
  function metricFromLaplacian(L) {
    var n = L.length, i, j, k;
    var e = eigSymmetric(L);
    var maxAbs = 0;
    for (i = 0; i < n; i++) maxAbs = Math.max(maxAbs, Math.abs(e.values[i]));
    // the same rank tolerance numpy/jax use for pinv on a symmetric matrix
    var tol = maxAbs * n * 2.220446049250313e-16;

    var P = [];
    for (i = 0; i < n; i++) P.push(new Array(n).fill(0));
    for (k = 0; k < n; k++) {
      if (Math.abs(e.values[k]) <= tol) continue;
      var inv = 1 / e.values[k];
      for (i = 0; i < n; i++) {
        var vik = e.vectors[i][k];
        if (vik === 0) continue;
        for (j = 0; j < n; j++) P[i][j] += inv * vik * e.vectors[j][k];
      }
    }
    var D = [];
    for (i = 0; i < n; i++) {
      D.push(new Array(n).fill(0));
      for (j = 0; j < n; j++) {
        var R = P[i][i] + P[j][j] - 2 * P[i][j];
        D[i][j] = Math.sqrt(R > 0 ? R : 0);
      }
    }
    return D;
  }

  /* Which nodes can actually reach which. pinv neither knows nor cares that a
   * graph is in pieces; this does. */
  function componentsOf(L, tol) {
    tol = tol || 1e-12;
    var n = L.length, label = new Array(n).fill(-1), next = 0, i, u, v;
    for (i = 0; i < n; i++) {
      if (label[i] !== -1) continue;
      var stack = [i];
      label[i] = next;
      while (stack.length) {
        u = stack.pop();
        for (v = 0; v < n; v++) {
          if (v !== u && Math.abs(L[u][v]) > tol && label[v] === -1) {
            label[v] = next;
            stack.push(v);
          }
        }
      }
      next++;
    }
    return label;
  }

  /* The distance matrix an interface may actually trust: the raw metric with
   * the two places pinv lies removed.
   *   - pairs in different pieces come back Infinity, not a small number
   *   - a mesh with no coupling left comes back all-Infinity, not all-zero */
  function meshMetric(L) {
    var n = L.length, i, j, off = 0, D;
    for (i = 0; i < n; i++) for (j = 0; j < n; j++) if (i !== j) off += Math.abs(L[i][j]);
    if (off < DEAD_MESH_EPS) {
      D = [];
      for (i = 0; i < n; i++) {
        D.push(new Array(n).fill(Infinity));
        D[i][i] = 0;
      }
      return { D: D, labels: Array.from({ length: n }, function (_, k) { return k; }), dead: true };
    }
    D = metricFromLaplacian(L);
    var lab = componentsOf(L);
    var pieces = Math.max.apply(null, lab);
    for (i = 0; i < n; i++) {
      for (j = 0; j < n; j++) {
        if (i === j) D[i][j] = 0;
        else if (pieces > 0 && lab[i] !== lab[j]) D[i][j] = Infinity;
      }
    }
    return { D: D, labels: lab, dead: false };
  }

  function laplacianFromEdges(n, edges) {
    var W = [], i;
    for (i = 0; i < n; i++) W.push(new Array(n).fill(0));
    edges.forEach(function (e) {
      if (e[0] === e[1]) throw new Error("a node cannot depend on itself: " + e[0]);
      W[e[0]][e[1]] = W[e[1]][e[0]] = e[2];
    });
    var L = [];
    for (i = 0; i < n; i++) {
      L.push(W[i].map(function (w) { return -w; }));
      L[i][i] = W[i].reduce(function (a, b) { return a + b; }, 0);
    }
    return L;
  }

  function ringLaplacian(n, J) {
    var edges = [], i;
    for (i = 0; i < n; i++) edges.push([i, (i + 1) % n, J]);
    return laplacianFromEdges(n, edges);
  }

  /* Classical MDS: double-centre the squared distances, take the top two
   * eigenvectors. The flat picture that best preserves the metric. */
  /* `axes` selects WHICH two eigenvectors, by rank. (0,1) is the classical
   * choice and keeps the most total structure. It is not always the one that
   * keeps the most useful structure -- see bestAxes. */
  function layout2d(D, keep, axes) {
    axes = axes || [0, 1];
    var m = keep.length, i, j;
    var B = [];
    for (i = 0; i < m; i++) B.push(new Array(m).fill(0));
    var sq = [];
    for (i = 0; i < m; i++) {
      sq.push(new Array(m).fill(0));
      for (j = 0; j < m; j++) {
        var d = D[keep[i]][keep[j]];
        sq[i][j] = d * d;
      }
    }
    var rowMean = sq.map(function (r) { return r.reduce(function (a, b) { return a + b; }, 0) / m; });
    var grand = rowMean.reduce(function (a, b) { return a + b; }, 0) / m;
    for (i = 0; i < m; i++) {
      for (j = 0; j < m; j++) B[i][j] = -0.5 * (sq[i][j] - rowMean[i] - rowMean[j] + grand);
    }
    var e = eigSymmetric(B);
    var order = e.values.map(function (v, k) { return [v, k]; })
      .sort(function (a, b) { return b[0] - a[0]; });
    var p1 = order[Math.min(axes[0], m - 1)], p2 = order[Math.min(axes[1], m - 1)];
    var out = [];
    for (i = 0; i < m; i++) {
      out.push([
        e.vectors[i][p1[1]] * Math.sqrt(Math.max(p1[0], 0)),
        e.vectors[i][p2[1]] * Math.sqrt(Math.max(p2[0], 0)),
      ]);
    }
    return out;
  }

  /* The worst pair in a layout, as a fraction of its true distance. 1.0 means
   * every distance on screen is the real one. */
  function flatness(D, keep, xy) {
    var worst = 1.0, a = -1, b = -1, i, j;
    for (i = 0; i < keep.length; i++) {
      for (j = i + 1; j < keep.length; j++) {
        var trueD = D[keep[i]][keep[j]];
        if (!isFinite(trueD) || trueD <= 0) continue;
        var ratio = Math.hypot(xy[i][0] - xy[j][0], xy[i][1] - xy[j][1]) / trueD;
        if (ratio < worst) { worst = ratio; a = keep[i]; b = keep[j]; }
      }
    }
    return { ratio: worst, a: a, b: b };
  }

  /* The plane that leaves NO pair collapsed, if there is one.
   *
   * Classical MDS minimises strain, which is a TOTAL -- and a total can be
   * excellent while one pair is destroyed. On the mesh SMI ships, the
   * strain-best plane draws two elements on top of each other (0% of their true
   * distance) while a different plane draws every pair at 71% or better.
   * Neither view is wrong; they answer different questions. */
  function bestAxes(D, keep, limit) {
    limit = Math.min(limit || 4, keep.length);
    var best = [0, 1], bestRatio = -1, i, j;
    for (i = 0; i < limit; i++) {
      for (j = i + 1; j < limit; j++) {
        var r = flatness(D, keep, layout2d(D, keep, [i, j])).ratio;
        if (r > bestRatio) { bestRatio = r; best = [i, j]; }
      }
    }
    return { axes: best, ratio: bestRatio };
  }

  /* ------------------------------------------------------- frame alignment --
   * Classical MDS fixes an embedding only up to rotation and reflection: the
   * eigenvectors are defined up to sign, so a small change in the graph can
   * hand back the same picture upside down. Deterministic per input, and to a
   * person dragging it, the map flipping under their finger reads as the
   * positions being arbitrary -- which is exactly the claim SMI makes against.
   *
   * So each new embedding is rotated/reflected onto the previous one, choosing
   * whichever orthogonal transform moves the shared nodes least. Real change
   * still shows; the cosmetic flips stop. This is the orthogonal Procrustes
   * problem, solved in closed form because 2-D needs no SVD.
   */
  function procrustes2d(Q, P) {
    var n = Math.min(Q.length, P.length), i;
    if (n < 2) return Q.slice();
    var qc = [0, 0], pc = [0, 0];
    for (i = 0; i < n; i++) {
      qc[0] += Q[i][0]; qc[1] += Q[i][1];
      pc[0] += P[i][0]; pc[1] += P[i][1];
    }
    qc = [qc[0] / n, qc[1] / n];
    pc = [pc[0] / n, pc[1] / n];

    function residual(theta, flip) {
      var c = Math.cos(theta), s2 = Math.sin(theta), tot = 0;
      for (var k = 0; k < n; k++) {
        var qx = (Q[k][0] - qc[0]) * (flip ? -1 : 1), qy = Q[k][1] - qc[1];
        var x = c * qx - s2 * qy, y = s2 * qx + c * qy;
        tot += (x - (P[k][0] - pc[0])) * (x - (P[k][0] - pc[0])) +
               (y - (P[k][1] - pc[1])) * (y - (P[k][1] - pc[1]));
      }
      return tot;
    }
    function bestAngle(flip) {
      var num = 0, den = 0;
      for (var k = 0; k < n; k++) {
        var qx = (Q[k][0] - qc[0]) * (flip ? -1 : 1), qy = Q[k][1] - qc[1];
        var px = P[k][0] - pc[0], py = P[k][1] - pc[1];
        num += qx * py - qy * px;
        den += qx * px + qy * py;
      }
      return Math.atan2(num, den);
    }
    var best = null;
    [false, true].forEach(function (flip) {
      var th = bestAngle(flip), r = residual(th, flip);
      if (best === null || r < best.r) best = { r: r, th: th, flip: flip };
    });
    var c = Math.cos(best.th), sn = Math.sin(best.th), out = [];
    for (i = 0; i < Q.length; i++) {
      var qx = (Q[i][0] - qc[0]) * (best.flip ? -1 : 1), qy = Q[i][1] - qc[1];
      out.push([c * qx - sn * qy + pc[0], sn * qx + c * qy + pc[1]]);
    }
    return out;
  }

  var API = {
    procrustes2d: procrustes2d,
    FADE_BELOW: FADE_BELOW,
    metricFromLaplacian: metricFromLaplacian,
    meshMetric: meshMetric,
    componentsOf: componentsOf,
    laplacianFromEdges: laplacianFromEdges,
    ringLaplacian: ringLaplacian,
    eigSymmetric: eigSymmetric,
    layout2d: layout2d,
    flatness: flatness,
    bestAxes: bestAxes,
    DEAD_MESH_EPS: DEAD_MESH_EPS,
  };

  if (typeof module !== "undefined" && module.exports) module.exports = API;
  else root.LMD = API;
})(typeof globalThis !== "undefined" ? globalThis : this);
